### PR TITLE
Populator event in vanilla compatibility generator

### PR DESCRIPTION
**Pull request by Foghrye4**
Here is a reason: https://minecraft.curseforge.com/projects/dungeon-generator-for-cubic-chunks?comment=5

### CODE REVIEW

**Comment by Niko-sk2x** [comment]
This is a significant change that was never intended to be done that way. This also touches on generating customcubic below y=0 in vanilla compatibility generator.

The problem is brings is that something that listens to this even may expect either a cubic chunks or a vanilla event, but not both. This is VERY likely to break mods that add CC compatibility and use this event.

**Comment by Foghrye4** [comment]
Some other special event then? Like `VanillaCompatibilityGeneratorPopulationEvent`?

**Comment by Niko-sk2x** [comment]
I don't really like that approach. Yet another event that specifically needs to be supported.

I think a good solution here would be to post CC events for < 0 and  > 256 only, and rely on vanilla events for everything else.

But this also has the problem that it will break CC generators that don't do generation in vanilla worlds.

The compatibility generator really should stay being just that: a compatibility generator. If you want to generate something in there, do it the vanilla way with all the limitations that follow.

**Comment by Niko-sk2x** [comment]
Another way would be to create a separate register method for the CC equivalent of IWorldGenerator. Something like this:

```java
public static void registerForCompatibilityGenerator(ICubicPopulator populator) {
   //code
}
```

The vanilla part can already be controlled with forge events, and there really isn't much to control outisde of the vanilla height range.

**Comment by Foghrye4** [comment]
Redone according your latest suggestion.

File: src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java [hunk 1]
```diff
@@ -323,4 +336,13 @@ public void recreateStructures(ICube cube) {
     public BlockPos getClosestStructure(String name, BlockPos pos, boolean findUnexplored) {
         return vanilla.getNearestStructurePos((World) world, name, pos, findUnexplored);
     }
+
+    /**
+     * Populators added here will be launched prior to any other. It is
+     * recommended to use this function in init or pre init event of a mod.
+     */
+    public static void registerForCompatibilityGenerator(ICubicPopulator populator) {
+        if (!customPopulators.contains(populator))
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java]: This needs to be in API.

**Comment by Foghrye4** [comment]
Moved to API.

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/CubeGeneratorsRegistry.java [hunk 2]
```diff
@@ -53,6 +53,9 @@
 
 public class CubeGeneratorsRegistry {
 
+    /** List of populators added by other mods to FlatCubic generator type */
+    public static final List<ICubicPopulator> customPopulatorsForFlatCubicGenerator = new ArrayList<ICubicPopulator>();
```
Comment (Niko-sk2x) [inline src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/CubeGeneratorsRegistry.java]: Don't make this public. Do it the same way it's done for normal populators. This list should not be directly accessible to mods.

**Comment by Foghrye4** [comment]
Privatized.

File: src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java [hunk 3]
```diff
@@ -261,6 +269,8 @@ public CubePrimer generateCube(int cubeX, int cubeY, int cubeZ) {
     @Override
     public void populate(ICube cube) {
         tryInit(vanilla, world);
+        Random rand = getChunkSpecificRandom(cube.getX(), cube.getZ());
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java]: Why not make the seed change with cube Y here too? It doesn't make sense to do for actual vanilla generator, but it does make sense for CC.

**Comment by Foghrye4** [comment]
Changed to use all three coordinates specific random

File: src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java [hunk 4]
```diff
@@ -183,16 +186,22 @@ public void generateColumn(Chunk column) {
...
 
     @Override
     public CubePrimer generateCube(int cubeX, int cubeY, int cubeZ) {
         tryInit(vanilla, world);
         CubePrimer primer = new CubePrimer();
 
         if (cubeY < 0) {
-            Random rand = new Random(world.getSeed());
-            rand.setSeed(rand.nextInt() ^ cubeX);
-            rand.setSeed(rand.nextInt() ^ cubeZ);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/core/worldgen/generator/vanilla/VanillaCompatibilityGenerator.java]: This should stay

**Pull request merged**